test/fixtures/alternative4-root-recordlit-let-success.dhall

```diff
@@ -0,0 +1,7 @@
+{ name = "my-project"
+, dependencies =
+      (let x = [ "console", "effect", "prelude", "psci-support" ] in x)
+    # [ "newtype" ]
```

Comment:
Well... I thought this would be the last test I would need to add, but stylistically, this is just stupid. It should be
```dhall
let x = [ "console", "effect", "newtype", "prelude", "psci-support" ] in x
```
However, that means the `x` could be a record where some part of it is selected:
```dhall
{ ...
, dependencies = let x = { deps = [ ... ] } in x.deps
}
```
Looks like I need to handle that possibility, too, by making all record expressions work even at the `WithinField` level.

Reply:
Basically, I would need to attempt to do the update within the let binding to account for the situation above. However, I would need to stop the traversal back out of the `WithinField` level if the expression ultimately refers to some other binding outside of that scope. For example:
```dhall
-- given
let lsBinding = ["old"] ... in { dependencies =  let x = lsBinding in x }

-- Wrong: side-effect introduced if `lsBinding` is used elsewhere besides the dependencies field
let lsBinding = ["old", "new"] ... in { dependencies = let x = lsBinding in x }

-- Correct: update occurs as close to local binding as possible
let lsBinding = ["old"] ... in { dependencies = let x = lsBinding # ["new"] in x  }
```

Reply:
I think for simplicity sake, let's not support this right now.

Let's say I have multiple let bindings here:
```dhall
let lsBinding = ["foo"]
{ ...
, dependencies = let x = lsBinding let y = x let z = y in z
}
```
When I hit the first let binding (i.e. `x`), I'm at level `WithinField`. When I attempt to update its `inExpr` (i.e. the `y` let binding), I'm still at level `WithinField`. When I finally get to the `z` let binding's `inExpr`, I now learn that it refers to the `y` binding's value. So what do I do?
1. Since I'm still at `WithinField`, do I just wrap the `z` binding's `inExpr` (e.g. `let ... let z = y # ["new"]`)? 
2. Or do I traverse back up the expression (e.g. `let x = lsBinding # ["new"] ... in z`)?

While the second is the obvious approach, the problem is that `WithinField` does not currently track how many let bindings we have gone down before we learned that we need to traverse back up the let binding expressions. Since it doesn't track that, the current code will eventually update the `lsBinding` binding's value rather than append the "new" package to the value of the `x` binding. To fix this, we'd need to update `WithinField` to also track how many let bindings it has gone down, so that if we would otherwise go back "up" the expression but find that the current counter is `0`, we know we must do the update here (i.e. at the `x` binding) rather than a binding that `x` also references (i.e. the `lsBinding`).

Since I've already written a lot of tests for this and don't feel like adding another 15 or so to account for this case, I think it's simpler to say, "We don't support it for now," because of how rare this will be. If we did implement this, we'd also need to account for things like field selection. For example,
```dhall
let lsBinding = {deps = ["foo"] }
{ ...
, dependencies = let x = lsBinding.deps let y = x let z = y in z
}
```

Reply:
Ok... I got curious to see how hard this would be, and was forced to do the "ExprLevel is just a stack of keys" optimization I mentioned in the original doc we covered this past Tuesday. In short, the latest commits make the amount of code I'm adding much smaller. Moreover, supporting targets is literally just adding a `"targets"` in front of the initial key stack we pass into the initial `updateExpr` call.

```diff
modifyRawDhallExpression :: HasLogFunc env => Text -> AstUpdate -> Expr -> RIO env Expr
modifyRawDhallExpression initialKey astMod originalExpr = do
-  result <- updateExpr (ExprLevel { levelKeyStack = ["dependencies" }) originalExpr
+  result <- updateExpr (ExprLevel { levelKeyStack = ["targets", "dependencies"] }) originalExpr
```

Reply:
This is wonderful!